comm/src/main/scala/coop/rchain/comm/transport/TcpTransportLayer.scala

```diff
@@ -51,51 +50,63 @@ class TcpTransportLayer[
         throw e
     }
 
-  private def clientChannel(remote: PeerNode): ManagedChannel =
-    NettyChannelBuilder
-      .forAddress(remote.endpoint.host, remote.endpoint.tcpPort)
-      .negotiationType(NegotiationType.TLS)
-      .sslContext(clientSslContext)
-      .intercept(new SslSessionClientInterceptor())
-      .overrideAuthority(remote.id.toString)
-      .build()
+  private def clientChannel(peer: PeerNode): Task[ManagedChannel] =
+    Task.delay {
+      NettyChannelBuilder
+        .forAddress(peer.endpoint.host, peer.endpoint.tcpPort)
+        .negotiationType(NegotiationType.TLS)
+        .sslContext(clientSslContext)
+        .intercept(new SslSessionClientInterceptor())
+        .overrideAuthority(peer.id.toString)
+        .build()
+    }
 
-  private def connection(remote: PeerNode): F[TransportLayerStub] =
+  private def connection(peer: PeerNode): Task[ManagedChannel] =
     for {
-      connections <- ConnectionsState[F].get
-      stub        = connections.getOrElse(remote.id, TransportLayerGrpc.stub(clientChannel(remote)))
-      _           <- ConnectionsState[F].modify(_ + (remote.id -> stub))
-    } yield stub
+      cs <- connections.get
+      c  <- cs.get(peer.id).fold(clientChannel(peer))(_.pure[Task])
+      _  <- connections.modify(_ + (peer.id -> c))
+    } yield c
 
-  def disconnect(remote: PeerNode): F[Unit] =
+  def disconnect(peer: PeerNode): Task[Unit] =
     for {
-      connections <- ConnectionsState[F].get
-      _ = connections
-        .get(remote.id)
-        .foreach(c => Try(c.getChannel.asInstanceOf[ManagedChannel].shutdown()))
-      _ <- ConnectionsState[F].modify(_ - remote.id)
+      cs <- connections.get
+      _ = cs
+        .get(peer.id)
+        .foreach(c => Try(c.shutdown()))
```

Comment:
foreach replace with fold
Try replace with Task

Reply:
Replaced with `match`. Is IMO more readable in this case.